dcc.py

```diff
@@ -372,20 +374,86 @@ def compile_dex(apkfile, filtercfg):
                 errors.append('%s:%s' % (full_name, str(e)))
                 continue
 
-            if code:
-                compiled_method_code[method_triple] = code
+            if code[0]:
+                compiled_method_code[method_triple] = code[0]
+                native_method_prototype[jni_longname] = code[1]
 
     return compiled_method_code, errors
 
 def is_apk(name):
     return name.endswith('.apk')
 
-def dcc_main(apkfile, filtercfg, outapk, do_compile=True, project_dir=None, source_archive='project-source.zip'):
+def write_empty_dynamic_register(project_dir, compiled_methods):
+    source_dir = os.path.join(project_dir, 'jni', 'nc')
+    if not os.path.exists(source_dir):
+        os.makedirs(source_dir)
+
+    filepath = os.path.join(source_dir, 'DynamicRegister.cpp')
+    with open(filepath, 'w') as fp:
+        fp.write('#include "DynamicRegister.h"\n\nconst char *DynamicRegister(JNIEnv *env) { return nullptr; }')
+
+def write_dynamic_register(project_dir, compiled_methods):
+    source_dir = os.path.join(project_dir, 'jni', 'nc')
+    if not os.path.exists(source_dir):
+        os.makedirs(source_dir)
+
+    exportList = {}
+
+    # Make export list
+    for method_triple, code in compiled_methods.items():
+        full_name = JniLongName(*method_triple)
+        if not full_name in native_method_prototype:
+            raise Exception('Method %s prototype info could not be found' % full_name)
+
+        class_path = method_triple[0][1:-1].replace('.', '/')
+        method_name = method_triple[1]
+        method_signature = method_triple[2]
+        method_native_name = full_name
+        method_native_prototype = native_method_prototype[full_name]
+
+        if not class_path in exportList:
+            exportList[class_path] = []
+        
+        exportList[class_path].append((method_name, method_signature, method_native_name, method_native_prototype))
+
+    if 0 == len(exportList):
+        logger.info('No export methods')
+        return
+    
+    # Generate extern block and export block
+    extern_block = ''
+    export_block = '\njclass clazz;\n\n'
+    export_block_count = 0
+    for class_path, methods in exportList.items():
```

Comment:
should sort exportList by key,the order of the items returned by dict.items() is not guaranteed,same environment, same input, should produce same output.

Reply:
you mean to make export_list and compiled_methods order the same?

Reply:
what we need is that if we compile a apk with same compile filter, we will get the same "DynamicRegister.cpp".
so we have to write:
 - extern entries
 - the order of JNINativeMethod elements
 - the order of RegisterNatives for compiled classes
in a deterministic way.
for example, we can do:
```
    for method_triple in sorted(compiled_methods.keys()):
        ...
```
```
    for class_path in sorted(exportList.keys()):
        methods = exportList[class_path] 
```